streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,623 @@
+            shards. Defaults to ``None``.
+        shuffle_seed (int, optional): Seed for shuffling, or ``None`` for random seed. Defaults to
+            ``None``.
+        shuffle_world_size (int, optional): Canonical world size for shuffling. Defaults to
+            ``None``, which is interpreted as the world size of the initial run.
+        batch_size (int, optional): Batch size of its DataLoader, which affects how the dataset is
+            partitioned over the workers. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 local: str,
+                 remote: Optional[str] = None,
+                 split: Optional[str] = None,
+                 shuffle: bool = False,
+                 predownload: Optional[int] = 100_000,
+                 keep_zip: Optional[bool] = None,
+                 download_retry: int = 2,
+                 download_timeout: float = 60,
+                 validate_hash: Optional[str] = None,
+                 shuffle_seed: Optional[int] = None,
+                 shuffle_world_size: Optional[int] = None,
+                 batch_size: Optional[int] = None):
+        self.local = local
+        self.remote = remote
+        self.split = split or ''  # Empty string for os.path.join().
+        self.shuffle = shuffle
+        self.predownload = predownload
+        self.keep_zip = keep_zip
+        self.download_retry = download_retry
+        self.download_timeout = download_timeout
+        self.validate_hash = validate_hash or None
+        # Seed is set below.
+        world = World()
+        if shuffle_world_size is None:
+            shuffle_world_size = world.num_workers
+        if shuffle_world_size < 1:
+            raise ValueError('Interleave must be at least one.')
+        self.shuffle_world_size = shuffle_world_size
+        self.batch_size = batch_size
+
+        # Load the index.json file.
+        basename = get_index_basename()
+        if world.is_local_leader:
+            filename = self._download_file(basename)
+        else:
+            filename = os.path.join(local, split, basename)  # pyright: ignore
+        dist.barrier()
+        obj = json.load(open(filename))
+        if obj['version'] != 2:
+            raise ValueError('Unsupported version')
+
+        # Initialize shard readers according to the loaded info.
+        self.shards = []
+        for info in obj['shards']:
+            shard = reader_from_json(local, split, info)
+            self.shards.append(shard)
+
+        # Build the Index (for partitioning and mapping samples to shards).
+        self.shard_sizes = np.array([x.samples for x in self.shards])
+        self.index = Index(self.shard_sizes)
+
+        # Determine and distribute shuffle seed and shm prefix.
+        if shuffle_seed is None:
+            shuffle_seed = np.random.randint(1 << 60)
+        prefix_int = np.random.randint(1 << 60)
+        if 1 < world.num_ranks:
```

Comment:
nit: can we do `world.num_ranks > 1`? a little more readable / consistent code style

Reply:
streaming repo has been ordering args LT < GT, let's future PR this